core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -122,46 +197,65 @@ public void setStoreListenerSource(Cache<K, V> source) {
+    CacheEventWrapper<K, V> cacheEventWrapper = new CacheEventWrapper<K, V>(event);
+    if (hasListeners()) {
+      threadLocal.verifyOrderedDispatch(orderedListenerCount.get() != 0);
+      threadLocal.addToEventList(cacheEventWrapper);
+      if (threadLocal.isOrdered()) {
+        ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue(event.getKey());
+        eventWrapperQueue.add(cacheEventWrapper);
       }
-      Runnable notificationTask = new Runnable() {
-        @Override
-        public void run() {
-          CacheEventListener<K, V> listener = wrapper.getListener();
-          listener.onEvent(event);
-        }
-      };
-      
-      ExecutorService eventDelivery = wrapper.config.orderingMode().equals(EventOrdering.UNORDERED) ? unorderedDelivery : orderedDelivery;
-      notificationResults.put(wrapper, eventDelivery.submit(notificationTask));
-    }
-    
-    for (Map.Entry<EventListenerWrapper, Future<?>> entry: notificationResults.entrySet()) {
-      EventListenerWrapper wrapper = entry.getKey();
-      Future<?> f = entry.getValue();
-      if ((EventFiring.SYNCHRONOUS.equals(wrapper.config.firingMode()))) {
-        boolean interrupted = false;
-        try {
-          f.get();
-        } catch (ExecutionException e) {
-            LOGGER.error("Cache Event Listener failed for event type {} due to ", type, e);
-          // XXX delegate to resilience strategy (#52), and/or just log
-        } catch (InterruptedException e) {
-          interrupted = true;
-        } finally {
-          if (interrupted) {
-            Thread.currentThread().interrupt();
-          }
+    }
+  }
+
+  private void dispatchEvent(CacheEventWrapper cacheEventWrapper, EventThreadLocal currentEventThreadLocal) {
+    cacheEventWrapper.markFireable();
+    ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue((K)cacheEventWrapper.cacheEvent.getKey());
+    orderedEventWait(cacheEventWrapper, eventWrapperQueue, currentEventThreadLocal);
+    if (currentEventThreadLocal.isOrdered()) {
+      while (!eventWrapperQueue.isEmpty() && eventWrapperQueue.peek().isFireable()) {
+      CacheEventWrapper event = eventWrapperQueue.remove();
+        actualDispatch(event, orderedEventDispatcher);
+        if (event != currentEventThreadLocal.get().get(0)) {
+          event.markFiredAndSignalCondition(true);
+        } else {
+          event.markFiredAndSignalCondition(false);
         }
       }
+    } else {
+      actualDispatch(cacheEventWrapper, unOrderedEventDispatcher);
+    }
+  }
+
+  private void orderedEventWait(CacheEventWrapper cacheEventWrapper, Queue<CacheEventWrapper> eventWrapperQueue, EventThreadLocal currentEventThreadLocal) {
+    if (currentEventThreadLocal.isOrdered()) {
+      if (cacheEventWrapper != eventWrapperQueue.peek()) {
+        cacheEventWrapper.waitTillFired();
+      }
+    }
+  }
+
+  private void actualDispatch(CacheEventWrapper event, EventDispatcher<K, V> eventDispatcher) {
```

Comment:
Objective: send event to the dispatcher for all sync/async listeners and facilitate synchronous dispatch

What it internally does:
•	If the event is not marked `FAILED`. It is sent to the eventDispatcher along with the set of syncListenerSet and then again with asyncListenerSet. 
•	The Future received form the syncListenerSet dispatcher is used to help with synchronous event firing

Why it should work:
In absence of sync listeners, threads would just come in, send their events to the eventDispatcher and continues without waiting for the event firing to be completed.
If synchronous listeners are present, threads that try to fire events will need to wait for the event firing to be completed. Future returned by the eventDispatcher aids in this regard.